tests/framework/prototype.py

```diff
@@ -41,39 +41,98 @@ def search_path(yaml_object, steps, default_value = None):
 # Functions for cluster preparation.
 def delete_kind_cluster() -> None:
     """Delete a KinD cluster"""
-    os.system("kind delete cluster")
+    shell_subprocess_run("kind delete cluster")
 
 def create_kind_cluster():
     """Create a KinD cluster"""
-    os.system("kind create cluster")
-    os.system("kubectl wait --for=condition=ready pod -n kube-system --all --timeout=900s")
+    shell_subprocess_run("kind create cluster")
+    shell_subprocess_run(
+        "kubectl wait --for=condition=ready pod -n kube-system --all --timeout=900s")
 
 def install_crd():
     """Install Custom Resource Definition (CRD)"""
-    os.system("kubectl create -k ../../manifests/cluster-scope-resources")
+    shell_subprocess_run("kubectl create -k ../../manifests/cluster-scope-resources")
 
 def download_images(docker_images):
     """Download Docker images from DockerHub"""
     docker_client = docker.from_env()
     for image in docker_images:
         # Only pull the image from DockerHub when the image does not
         # exist in the local docker registry.
-        if os.system(f'docker image inspect {image} > /dev/null') != 0:
+        if shell_subprocess_run(f'docker image inspect {image} > /dev/null', check = False) != 0:
             docker_client.images.pull(image)
     docker_client.close()
 
 def kind_load_images(docker_images):
     """Load downloaded images into KinD cluster"""
     for image in docker_images:
-        os.system(f'kind load docker-image {image}')
+        shell_subprocess_run(f'kind load docker-image {image}')
 
 def install_kuberay_operator():
     """Install kuberay operator with image kuberay/operator:nightly"""
-    os.system('kubectl apply -k ../../manifests/base')
+    shell_subprocess_run('kubectl apply -k ../../manifests/base')
 
 def check_cluster_exist():
     """Check whether KinD cluster exists or not"""
-    return os.system("kubectl cluster-info --context kind-kind") == 0
+    return shell_subprocess_run("kubectl cluster-info --context kind-kind", check = False) == 0
+
+def check_pod_running(pods) -> bool:
+    """"Check whether all of the pods are in running state"""
+    for pod in pods:
+        if pod.status.phase != 'Running':
+            return False
+    return True
+
+def shell_subprocess_run(command, check = True):
+    """Run the command using subprocess.run() with shell=True"""
+    return subprocess.run(command, shell = True, check = check).returncode
+
+def get_expected_head_pods(custom_resource):
```

Comment:
```python
def get_expected_head_pods(custom_resource):
    """Get the number of head pods in custom_resource"""
    resource_kind = custom_resource["kind"]
    head_replica_paths = {
       "RayCluster": "spec.headGroupSpec.replicas",
       "RayService": "spec.rayClusterConfig.headGroupSpec.replicas",
       "RayJob": "spec.rayClusterSpec.headGroupSpec.replicas"
    }
    if resource_kind in head_replica_paths:
       path = head_replica_paths[resource_kind]
       return search_path(custom_resource, path.split('.'), default_value=1)
    raise Exception(f"Unknown resource kind: {resource_kind} in get_expected_head_pods()")
```

In the future, we can move `head_replicas_paths` to an independent module "constant". Feel free to rename the variables.